src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -62,20 +62,41 @@ internal class PackageRestoreDataSource : ChainedProjectValueDataSourceBase<Rest
         private readonly IFileSystem _fileSystem;
         private readonly IProjectDiagnosticOutputService _logger;
         private readonly IProjectDependentFileChangeNotificationService _projectDependentFileChangeNotificationService;
-
+        private readonly IVsSolutionRestoreService4 _solutionRestoreService4;
         private byte[]? _latestHash;
         private bool _enabled;
 
+        /// <summary>
+        /// Project Unique Name used by Nuget Nomination.
+        /// </summary>
+        public string Name => _project.FullPath;
+
+        /// <summary>
+        /// Re-usable task that completes when there is a new nomination
+        /// </summary>
+        private TaskCompletionSource<bool>? _whenNominatedTask;
+
+        /// <summary>
+        /// Save the configured project versions that might get nominations.
+        /// </summary>
+        private readonly Dictionary<ProjectConfiguration, IComparable> _savedNominatedVersion = new();
+
+        /// <summary>
+        /// Object used to synchronize access to IVsProjectRestoreInfoSource.
+        /// </summary>
+        private readonly object _object = new();
+
         [ImportingConstructor]
         public PackageRestoreDataSource(
             UnconfiguredProject project,
             IPackageRestoreUnconfiguredInputDataSource dataSource,
-            [Import(ExportContractNames.Scopes.UnconfiguredProject)]IProjectAsynchronousTasksService projectAsynchronousTasksService,
+            [Import(ExportContractNames.Scopes.UnconfiguredProject)] IProjectAsynchronousTasksService projectAsynchronousTasksService,
             IVsSolutionRestoreService3 solutionRestoreService,
             IFileSystem fileSystem,
             IProjectDiagnosticOutputService logger,
-            IProjectDependentFileChangeNotificationService projectDependentFileChangeNotificationService)
-            : base(project, synchronousDisposal : true, registerDataSource : false)
+            IProjectDependentFileChangeNotificationService projectDependentFileChangeNotificationService,
+            IVsSolutionRestoreService4 solutionRestoreService4)
+            : base(project, synchronousDisposal: true, registerDataSource: false)
```

Comment:
I realize that a new problem with this PR is that we might introduce new deadlock point. When a feature is blocking on UI thread to wait NuGet restore to finish, (a scenario like this is that a feature blocking UI to wait intellisense ready for a specific project), it now waits all project in the solution to provide NuGet restore info. And that process might require UI thread (a scenario like this is that one of the project in the solution might need UI thread by the host object to run code generation in its design time build -- a possibility if WPF is enabled in one of the project.

I think a relative easy way to fix it is to get all PackageRestoreDataSource to share a same JTF collection/factory, which might require a change to the base class in CPS to enable it.  I plan to create a CPS PR to do.

Reply:
I added a new constructor to pass in IJoinableTaskScope (which was defined to prevent threading validation rules to be triggered if we pass in JoinableTaskCollection/Factory directly.) To prevent this to be used by accident, I marked this new constructor advanced, which will not be shown in a part of basic intelisense. I guess creating a ProjectService scope component to implement that, and import that component in the constructor of PackageRestoreDataSource will allow all of them to share the same collection, which will address this issue without changing the other part of code.

Reply:
[CPS PR](https://devdiv.visualstudio.com/DevDiv/_git/CPS/pullrequest/337260)

Reply:
"I guess creating a ProjectService scope component to implement that, and import that component in the constructor of PackageRestoreDataSource"

Do you mean create a new `IProjectServices `and import it in `PackageRestoreDataSource(SharedJoinableTaskCollection c)` ?
```
[Export(typeof(SharedJoinableTaskCollection ))]
class SharedJoinableTaskCollection : IProjectServices {

    IJoinableTaskScope joinableTaskScope;

   SharedJoinableTaskCollection()
   {
      joinableTaskScope.JoinableTaskCollection = joinableTaskContext.CreateCollection();
      joinableTaskScope.JoinableTaskFactory =  joinableTaskContext.CreateDefaultFactory();
   }

}
```